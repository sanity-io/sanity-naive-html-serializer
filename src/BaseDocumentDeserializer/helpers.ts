import {htmlToBlocks} from '@sanity/block-tools'
import {Schema} from '@sanity/schema'
import {ObjectField, PortableTextSpan, PortableTextTextBlock} from 'sanity'

const defaultSchema = Schema.compile({
  name: 'default',
  types: [
    {
      type: 'object',
      name: 'default',
      fields: [
        {
          name: 'block',
          type: 'array',
          of: [{type: 'block'}],
        },
      ],
    },
  ],
})

export const blockContentType = defaultSchema
  .get('default')
  .fields.find((field: ObjectField) => field.name === 'block').type

export const noSchemaWarning = (obj: Element): string =>
  `WARNING: Unfortunately the deserializer may have issues with this field or object: ${obj.className}.
  If it's a specific type, you may need to declare  at the top level, or write a custom deserializer.`

//helper to handle messy input -- take advantage
//of blockTools' sanitizing behavior for single strings
export const preprocess = (html: string): string => {
  const intermediateBlocks = htmlToBlocks(
    `<p>${html}</p>`,
    blockContentType
  ) as PortableTextTextBlock<PortableTextSpan>[]
  if (!intermediateBlocks.length) {
    throw new Error(`Error parsing string '${html}'`)
  }
  return intermediateBlocks[0].children[0].text
}
