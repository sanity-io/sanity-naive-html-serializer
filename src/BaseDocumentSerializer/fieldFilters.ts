import { ObjectField, TypedObject } from '@sanity/types'

const META_FIELDS = ['_key', '_type', '_id']

/*
 * Helper. If field-level translation pattern used, only sends over
 * content from the base language. Works recursively, so if users
 * use this pattern several layers deep, base language fields will still be found.
 */
export const languageObjectFieldFilter = (
  obj: Record<string, any>,
  baseLang: string
) => {
  const filterToLangField = (obj: Record<string, any>) => {
    const filteredObj: Record<string, any> = {}
    filteredObj[baseLang] = obj[baseLang]
    META_FIELDS.forEach(field => {
      if (obj[field]) {
        filteredObj[field] = obj[field]
      }
    })
    return filteredObj
  }

  const findBaseLang = (childObj: Record<string, any>): Record<string, any> => {
    const filteredObj: Record<string, any> = {}
    META_FIELDS.forEach(field => {
      if (childObj[field]) {
        filteredObj[field] = childObj[field]
      }
    })

    for (let key in childObj) {
      const value: any = childObj[key]
      //we've reached a base language field, add it to
      //what we want to send to translation
      if (value.hasOwnProperty(baseLang)) {
        filteredObj[key] = filterToLangField(value)
      }
      //we have an array that may have language fields in its objects
      else if (
        Array.isArray(value) &&
        value.length &&
        typeof value[0] === 'object'
      ) {
        //recursively find and filter for any objects that have the base language
        const validLangObjects = value.reduce((validArr, objInArray) => {
          if (objInArray._type === 'block') {
            validArr.push(objInArray)
          } else if (objInArray.hasOwnProperty(baseLang)) {
            validArr.push(filterToLangField(objInArray))
          } else {
            const filtered = findBaseLang(objInArray)
            if (Object.keys(filtered).length) {
              validArr.push(filtered)
            }
          }
          return validArr
        }, [])
        if (validLangObjects.length) {
          filteredObj[key] = validLangObjects
        }
      }
      //we have an object nested in an object
      //recurse down the tree
      else if (typeof value === 'object') {
        const nestedLangObj = findBaseLang(value)
        if (Object.keys(nestedLangObj).length) {
          filteredObj[key] = nestedLangObj
        }
      }
    }
    return filteredObj
  }

  //send top level object into recursive function
  return findBaseLang(obj)
}

/*
 * Eliminates stop-types and non-localizable fields
 * for document-level translation.
 */
export const fieldFilter = (
  obj: Record<string, any>,
  objFields: ObjectField[],
  stopTypes: string[]
) => {
  const filteredObj: TypedObject = { _type: obj._type }

  const fieldFilter = (field: Record<string, any>) => {
    if (field.localize === false) {
      return false
    } else if (field.type === 'string' || field.type === 'text') {
      return true
    } else if (Array.isArray(obj[field.name])) {
      return true
    } else if (!stopTypes.includes(field.type)) {
      return true
    }
    return false
  }

  const validFields = [
    ...META_FIELDS,
    ...objFields.filter(fieldFilter).map(field => field.name),
  ]
  validFields.forEach(field => {
    if (obj[field]) {
      filteredObj[field] = obj[field]
    }
  })
  return filteredObj
}
