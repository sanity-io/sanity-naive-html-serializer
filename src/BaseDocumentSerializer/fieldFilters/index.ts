export {fieldFilter} from './baseFieldFilter'
export {languageObjectFieldFilter} from './languageObjectFieldFilter'
export {internationalizedArrayFilter} from './internationalizedArrayFilter'
