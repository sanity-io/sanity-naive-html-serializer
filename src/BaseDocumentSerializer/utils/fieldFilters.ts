import {ObjectField, TypedObject} from 'sanity'

const META_FIELDS = ['_key', '_type', '_id']

/*
 * Helper. If field-level translation pattern used, only sends over
 * content from the base language. Works recursively, so if users
 * use this pattern several layers deep, base language fields will still be found.
 */
export const languageObjectFieldFilter = (
  obj: Record<string, any>,
  baseLang: string
): Record<string, any> => {
  const filterToLangField = (childObj: Record<string, any>) => {
    const filteredObj: Record<string, any> = {}
    filteredObj[baseLang] = childObj[baseLang]
    META_FIELDS.forEach((field) => {
      if (childObj[field]) {
        filteredObj[field] = childObj[field]
      }
    })
    return filteredObj
  }

  const findBaseLang = (childObj: Record<string, any>): Record<string, any> => {
    const filteredObj: Record<string, any> = {}
    META_FIELDS.forEach((field) => {
      if (childObj[field]) {
        filteredObj[field] = childObj[field]
      }
    })

    for (const key in childObj) {
      if (childObj.hasOwnProperty(key)) {
        const value: any = childObj[key]
        //we've reached a base language field, add it to
        //what we want to send to translation
        if (value.hasOwnProperty(baseLang)) {
          filteredObj[key] = filterToLangField(value)
        }
        //we have an array that may have language fields in its objects
        else if (Array.isArray(value) && value.length && typeof value[0] === 'object') {
          //recursively find and filter for any objects that have the base language
          const validLangObjects = value.reduce((validArr, objInArray) => {
            if (objInArray._type === 'block') {
              validArr.push(objInArray)
            } else if (objInArray.hasOwnProperty(baseLang)) {
              validArr.push(filterToLangField(objInArray))
            } else {
              const filtered = findBaseLang(objInArray)
              const nonMetaFields = Object.keys(filtered).filter(
                (objInArrayKey) => META_FIELDS.indexOf(objInArrayKey) === -1
              )
              if (nonMetaFields.length) {
                validArr.push(filtered)
              }
            }
            return validArr
          }, [])
          if (validLangObjects.length) {
            filteredObj[key] = validLangObjects
          }
        }
        //we have an object nested in an object
        //recurse down the tree
        else if (typeof value === 'object') {
          const nestedLangObj = findBaseLang(value)
          const nonMetaFields = Object.keys(nestedLangObj).filter(
            (nestedObjKey) => META_FIELDS.indexOf(nestedObjKey) === -1
          )
          if (nonMetaFields.length) {
            filteredObj[key] = nestedLangObj
          }
        }
      }
    }
    return filteredObj
  }

  //send top level object into recursive function
  return findBaseLang(obj)
}

/*
 * Eliminates stop-types and non-localizable fields
 * for document-level translation.
 */
export const fieldFilter = (
  obj: Record<string, any>,
  objFields: ObjectField[],
  stopTypes: string[]
): TypedObject => {
  const filteredObj: TypedObject = {_type: obj._type}

  const fieldFilterFunc = (field: Record<string, any>) => {
    if (field.localize === false) {
      return false
    } else if (field.type === 'string' || field.type === 'text') {
      return true
    } else if (Array.isArray(obj[field.name])) {
      return true
    } else if (!stopTypes.includes(field.type)) {
      return true
    }
    return false
  }

  const validFields = [
    ...META_FIELDS,
    ...objFields?.filter(fieldFilterFunc)?.map((field) => field.name),
  ]
  validFields.forEach((field) => {
    if (obj[field]) {
      filteredObj[field] = obj[field]
    }
  })
  return filteredObj
}
