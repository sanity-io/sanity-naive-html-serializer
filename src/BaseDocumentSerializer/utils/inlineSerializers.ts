import {PortableTextComponents} from '@portabletext/to-html'
import {ArbitraryTypedObject} from '@portabletext/types'
import {PortableTextSpan, PortableTextTextBlock, Schema} from 'sanity'
import {BaseDocumentSerializer} from '..'

type ObjectSerializerMethod = ({value}: {value: ArbitraryTypedObject}) => string
type MarkSerializerMethod = ({
  value,
  markType,
  children,
}: {
  value: ArbitraryTypedObject
  markType: string
  children: any[]
}) => string

const validateInlineContent = (
  obj: ArbitraryTypedObject | PortableTextSpan,
  serializers: PortableTextComponents,
  stopTypes: string[]
) =>
  //don't serialize blocks or spans
  obj._type !== 'block' &&
  obj._type !== 'span' &&
  //don't serialize any types that have been explicitly marked as stop types
  !stopTypes.includes(obj._type) &&
  //don't override custom serialization methods from the user
  !serializers.types?.[obj._type]

export const createInlineSerializerTypes = (
  block: PortableTextTextBlock,
  stopTypes: string[],
  serializers: PortableTextComponents,
  schemas: Schema
): PortableTextComponents => {
  const inlineObjects = block.children.filter((child) =>
    validateInlineContent(child, serializers, stopTypes)
  )
  const types: string[] = Array.from(new Set(inlineObjects.map((obj) => obj._type)))
  const serializerFunc: ObjectSerializerMethod = ({value}: {value: ArbitraryTypedObject}) => {
    const wrapperStart = [
      `<span class="default-serialized-object"`,
      `id="${value._key ?? value._id}"`,
      `data-type="${value._type}">`,
    ].join(' ')
    const serializedData = BaseDocumentSerializer(schemas).serializeObject(
      value,
      stopTypes,
      serializers
    )
    const wrapperEnd = `</span>`
    return [wrapperStart, serializedData, wrapperEnd].join('')
  }
  return types.reduce<Record<string, ObjectSerializerMethod>>((acc, type) => {
    acc[type as string] = serializerFunc
    return acc
  }, {})
}

export const createMarkSerializerTypes = (
  block: PortableTextTextBlock,
  stopTypes: string[],
  serializers: PortableTextComponents,
  schemas: Schema
): PortableTextComponents => {
  const marks = block.markDefs?.filter((mark) =>
    validateInlineContent(mark, serializers, stopTypes)
  )
  const markTypes = Array.from(new Set(marks?.map((mark) => mark._type)))
  const serializerFunc = ({
    value,
    markType,
    children,
  }: {
    value: ArbitraryTypedObject
    markType: string
    children: any[]
  }) => {
    const wrapperStart = [
      `<span class="default-serialized-annotation"`,
      `id="${value._key ?? value._id}"`,
      `data-type="${markType}">`,
    ].join(' ')
    const serializedChildren = `<span class="default-serialized-children">${children}</span>`
    const serializedData = BaseDocumentSerializer(schemas).serializeObject(
      value,
      stopTypes,
      serializers
    )
    const wrapperEnd = `</span>`
    return [wrapperStart, serializedChildren, serializedData, wrapperEnd].join('')
  }
  return markTypes.reduce<Record<string, MarkSerializerMethod>>((acc, type) => {
    acc[type as string] = serializerFunc
    return acc
  }, {})
}
