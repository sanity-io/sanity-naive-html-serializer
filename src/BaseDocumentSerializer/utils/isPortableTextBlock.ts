import {ArbitraryTypedObject} from '@portabletext/types'
import {PortableTextTextBlock} from 'sanity'

export const isPortableTextBlock = (el: ArbitraryTypedObject): el is PortableTextTextBlock => {
  return el._type === 'block'
}
