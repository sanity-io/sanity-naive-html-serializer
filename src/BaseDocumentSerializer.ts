import schemas from 'part:@sanity/base/schema'
import blocksToHtml, { h } from '@sanity/block-content-to-html'
import { defaultStopTypes, customSerializers } from './BaseSerializationConfig'
import { ObjectField } from '@sanity/types'
import { findLatestDraft } from './helpers'
import { Serializer } from './types'
import clone from 'just-clone'

/*
 * Helper function that allows us to get metadata (like `localize: false`) from schema fields.
 */
const getSchema = (name: string) =>
  schemas._original.types.find((s: ObjectField) => s.name === name)

/*
 * Main parent function: finds fields to translate, and feeds them to appropriate child serialization
 * methods.
 */
const serializeDocument = async (
  documentId: string,
  translationLevel: string = 'document',
  baseLang: string = 'en',
  stopTypes: string[] = defaultStopTypes,
  serializers: Record<string, any> = customSerializers
) => {
  const doc = await findLatestDraft(documentId)

  let filteredObj: Record<string, any> = {}

  if (translationLevel === 'field') {
    filteredObj = languageObjectFieldFilter(doc, baseLang)
  } else {
    filteredObj = fieldFilter(doc, getSchema(doc._type).fields, stopTypes)
  }

  const serializedFields: Record<string, any> = {}
  for (let key in filteredObj) {
    const value: Record<string, any> | Array<any> = filteredObj[key]
    if (typeof value === 'string') {
      serializedFields[key] = value
    } else if (Array.isArray(value)) {
      serializedFields[key] = serializeArray(value, key, stopTypes, serializers)
    } else {
      serializedFields[key] = serializeObject(
        value,
        key,
        stopTypes,
        serializers
      )
    }
  }

  return {
    name: documentId,
    content: serializeObject(
      serializedFields,
      doc._type,
      stopTypes,
      serializers
    ),
  }
}

/*
 * Helper. If field-level translation pattern used, only sends over
 * content from the base language.
 */
const languageObjectFieldFilter = (
  obj: Record<string, any>,
  baseLang: string
) => {
  const filteredObj: Record<string, any> = {}
  for (let key in obj) {
    const value: any = obj[key]
    if (value.hasOwnProperty(baseLang)) {
      filteredObj[key] = {}
      filteredObj[key][baseLang] = value[baseLang]
    }
  }

  return filteredObj
}

/*
 * Helper. Eliminates stop-types and non-localizable fields.
 */
const fieldFilter = (
  obj: Record<string, any>,
  objFields: ObjectField[],
  stopTypes: string[]
) => {
  const filteredObj: Record<string, any> = {}

  const fieldFilter = (field: Record<string, any>) => {
    if (field.localize === false) {
      return false
    } else if (field.type.name === 'string' || field.type.name === 'text') {
      return true
    } else if (Array.isArray(obj[field.name])) {
      return true
    } else if (!stopTypes.includes(field.type.name)) {
      return true
    }
    return false
  }

  const validFields = [
    '_key',
    '_type',
    '_id',
    ...objFields.filter(fieldFilter).map(field => field.name),
  ]
  validFields.forEach(field => {
    if (obj[field]) {
      filteredObj[field] = obj[field]
    }
  })
  return filteredObj
}

const serializeArray = (
  fieldContent: Record<string, any>[],
  fieldName: string,
  stopTypes: string[],
  serializers: Record<string, any>
) => {
  const validBlocks = fieldContent.filter(
    block => !stopTypes.includes(block._type)
  )

  const filteredBlocks = validBlocks.map(block => {
    const schema = getSchema(block._type)
    if (schema) {
      return fieldFilter(block, schema.fields, stopTypes)
    } else {
      return block
    }
  })

  const output = filteredBlocks.map(obj =>
    //deep copy needed here to avoid serializer rules
    //serializing objects with content that comes before in the loop
    serializeObject(obj, null, stopTypes, serializers)
  )

  return `<div class="${fieldName}">${output.join('')}</div>`
}

const serializeObject = (
  obj: Record<string, any>,
  topFieldName: string | null = null,
  stopTypes: string[],
  serializers: Record<string, any>
) => {
  if (stopTypes.includes(obj._type)) {
    return ''
  }

  const hasSerializer =
    serializers.types && Object.keys(serializers.types).includes(obj._type)
  if (hasSerializer) {
    return blocksToHtml({ blocks: [obj], serializers: serializers })
  }

  const tempSerializers = clone(serializers)

  if (obj._type !== 'span' && obj._type !== 'block') {
    let innerHTML = ''
    Object.entries(obj).forEach(([fieldName, value]) => {
      let htmlField = ''

      if (!['_key', '_type', '_id'].includes(fieldName)) {
        if (typeof value === 'string') {
          const htmlRegex = /^</
          //this field may have been recursively turned into html already.
          htmlField = value.match(htmlRegex)
            ? value
            : `<span class="${fieldName}">${value}</span>`
        } else if (Array.isArray(value)) {
          htmlField = serializeArray(value, fieldName, stopTypes, serializers)
        } else {
          const schema = getSchema(value._type)
          let toTranslate = value
          if (schema) {
            toTranslate = fieldFilter(value, schema.fields, stopTypes)
          }
          const objHTML = serializeObject(
            toTranslate,
            null,
            stopTypes,
            serializers
          )
          htmlField = `<div class=${fieldName}>${objHTML}</div>`
        }
      }
      innerHTML += htmlField
    })

    if (!innerHTML) {
      return ''
    }
    tempSerializers.types[obj._type] = (props: Record<string, any>) => {
      return h('div', {
        className: topFieldName ?? props.node._type,
        id: props.node._key ?? props.node._id,
        innerHTML: innerHTML,
      })
    }
  }

  let serializedBlock = ''
  try {
    serializedBlock = blocksToHtml({
      blocks: [obj],
      serializers: tempSerializers,
    })
  } catch (err) {
    console.debug(
      `Had issues serializing block of type "${obj._type}". Please specify a serialization method for this block in your serialization config. Received error: ${err}`
    )
  }

  return serializedBlock
}

export const BaseDocumentSerializer: Serializer = {
  serializeDocument,
  fieldFilter,
  languageObjectFieldFilter,
  serializeArray,
  serializeObject,
}
