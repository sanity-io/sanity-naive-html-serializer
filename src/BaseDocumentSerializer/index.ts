import {defaultStopTypes, customSerializers} from '../BaseSerializationConfig'
import {SanityDocument, TypedObject, Schema} from 'sanity'
import {TranslationLevel, SerializerClosure} from '../types'
import {fieldFilter, languageObjectFieldFilter} from './fieldFilters'
import {PortableTextTypeComponent, toHTML} from '@portabletext/to-html'

const META_FIELDS = ['_key', '_type', '_id', '_weak']

export const BaseDocumentSerializer: SerializerClosure = (schemas: Schema) => {
  /*
   * Helper function that allows us to get metadata (like `localize: false`) from schema fields.
   */
  const getSchema = (name: string) => schemas?._original?.types.find((s) => s.name === name) as any

  const serializeObject = (
    obj: TypedObject,
    stopTypes: string[],
    serializers: Record<string, any>
  ) => {
    if (stopTypes.includes(obj._type)) {
      return ''
    }

    //if user has declared a custom serializer, use that
    //instead of this method
    const hasSerializer = serializers.types && Object.keys(serializers.types).includes(obj._type)
    if (hasSerializer) {
      return toHTML([obj], {components: serializers})
    }

    //we don't need to worry about PT types
    if (obj._type === 'span' || obj._type === 'block') {
      return toHTML(obj, {components: serializers})
    }

    //if schema is available, encode values in the order they're declared in the schema,
    //since this will likely be more intuitive for a translator.
    let fieldNames = Object.keys(obj).filter((key) => key !== '_type')
    const schema = getSchema(obj._type)
    if (schema && schema.fields) {
      fieldNames = schema.fields
        .map((field: Record<string, any>) => field.name)
        .filter((schemaKey: string) => Object.keys(obj).includes(schemaKey))
    }

    //account for anonymous inline objects
    if (typeof obj === 'object' && !obj._type) {
      obj._type = ''
    }

    //in some cases, we might recurse through many objects of the same type
    //we should take all methods necessary to ensure state does not persist
    //otherwise we risk using old serialization methods on new items
    const newSerializationMethods: Record<string, PortableTextTypeComponent> = {}
    const tempType = `${obj._type}__temp_type__${Math.random().toString(36).substring(7)}`
    const objToSerialize: TypedObject = {_type: tempType}
    //for our default serialization method, we only need to
    //capture metadata. the rest will be recursively turned into strings.
    META_FIELDS.filter((f) => f !== '_type').forEach((field) => {
      objToSerialize[field] = obj[field]
    })

    let innerHTML = ''

    //if it's a custom object, iterate through its keys to find and serialize translatable content
    fieldNames.forEach((fieldName) => {
      let htmlField = ''

      if (!META_FIELDS.includes(fieldName)) {
        const value = obj[fieldName]
        //strings are either string fields or have recursively been turned
        //into HTML because they were a nested object or array
        if (typeof value === 'string') {
          const htmlRegex = new RegExp(/<("[^"]*"|'[^']*'|[^'">])*>/)
          if (htmlRegex.test(value)) {
            htmlField = value
          } else {
            htmlField = `<span class="${fieldName}">${value}</span>`
          }
        }

        //array fields get filtered and its children serialized
        else if (Array.isArray(value)) {
          //eslint-disable-next-line no-use-before-define -- this is a recursive function
          htmlField = serializeArray(value, fieldName, stopTypes, {
            ...serializers,
            types: {...serializers.types},
          })
        }

        //this is an object in an object, serialize it first
        else {
          const embeddedObject = value as TypedObject
          const embeddedObjectSchema = getSchema(embeddedObject._type)
          let toTranslate = embeddedObject
          if (embeddedObjectSchema && embeddedObjectSchema.fields) {
            toTranslate = fieldFilter(toTranslate, embeddedObjectSchema.fields, stopTypes)
          }
          const objHTML = serializeObject(toTranslate, stopTypes, {
            ...serializers,
            types: {...serializers.types},
          })
          htmlField = `<div class="${fieldName}" data-level="field">${objHTML}</div>`
        }

        innerHTML += htmlField
      }
    })

    if (!innerHTML) {
      return ''
    }

    newSerializationMethods[tempType] = ({value}: {value: TypedObject}) => {
      let div = `<div class="${value._type.split('__temp_type__')[0]}"`
      if (value._key || value._id) {
        div += `id="${value._key ?? value._id}"`
      }

      return [div, ` data-type="object">${innerHTML}</div>`].join('')
    }

    let serializedBlock = ''
    try {
      serializedBlock = toHTML(objToSerialize, {
        components: {
          ...serializers,
          types: {
            ...serializers.types,
            ...newSerializationMethods,
          },
        },
      })
    } catch (err) {
      //eslint-disable-next-line no-console -- this is a warning
      console.warn(
        `Had issues serializing block of type "${obj._type}". Please specify a serialization method for this block in your serialization config. Received error: ${err}`
      )
    }

    return serializedBlock
  }

  const serializeArray = (
    fieldContent: Record<string, any>[],
    fieldName: string,
    stopTypes: string[],
    serializers: Record<string, any>
  ) => {
    //filter for any blocks that user has indicated
    //should not be sent for translation
    const validBlocks = fieldContent.filter((block) => !stopTypes.includes(block._type))

    //take out any fields in these blocks that should
    //not be sent to translation
    const filteredBlocks = validBlocks.map((block) => {
      const schema = getSchema(block._type)
      if (schema && schema.fields) {
        return fieldFilter(block, schema.fields, stopTypes)
      }
      return block
    })

    const output = filteredBlocks.map((obj) => {
      //if object in array is just a string, just return it
      if (typeof obj === 'string') {
        return `<span>${obj}</span>`
      }
      //send to serialization method
      return serializeObject(obj as TypedObject, stopTypes, serializers)
    })

    //encode this with data-level field
    return `<div class="${fieldName}" data-type="array">${output.join('')}</div>`
  }

  /*
   * Main parent function: finds fields to translate, and feeds them to appropriate child serialization
   * methods.
   */
  const serializeDocument = (
    doc: SanityDocument,
    translationLevel: TranslationLevel = 'document',
    baseLang = 'en',
    stopTypes = defaultStopTypes,
    serializers = customSerializers
  ) => {
    const schema = getSchema(doc._type)
    let filteredObj: Record<string, any> = {}

    //field level translations explicitly send over any fields that
    //match the base language, regardless of depth
    if (translationLevel === 'field') {
      filteredObj = languageObjectFieldFilter(doc, baseLang)
    }
    //otherwise, we can refer to the schema and a list of stop types
    //to determine what should not be sent
    else {
      filteredObj = fieldFilter(doc, schema.fields, stopTypes)
    }

    const serializedFields: Record<string, any> = {}

    for (const key in filteredObj) {
      if (filteredObj.hasOwnProperty(key) === false) continue
      const value: Record<string, any> | Array<any> | string = filteredObj[key]

      if (typeof value === 'string') {
        serializedFields[key] = value
      } else if (Array.isArray(value)) {
        serializedFields[key] = serializeArray(value, key, stopTypes, serializers)
      } else if (value && !stopTypes.find((stopType) => stopType == value?._type)) {
        const serialized = serializeObject(value as TypedObject, stopTypes, serializers)
        serializedFields[key] = `<div class="${key}" data-level='field'>${serialized}</div>`
      }
    }

    //create a valid HTML file
    const rawHTMLBody = document.createElement('body')
    rawHTMLBody.innerHTML = serializeObject(serializedFields as TypedObject, stopTypes, serializers)

    const rawHTMLHead = document.createElement('head')
    const metaFields = ['_id', '_type', '_rev']
    //save our metadata as meta tags so we can use them later on
    metaFields.forEach((field) => {
      const metaEl = document.createElement('meta')
      metaEl.setAttribute('name', field)
      metaEl.setAttribute('content', doc[field] as string)
      rawHTMLHead.appendChild(metaEl)
    })
    //encode version so we can use the correct deserialization methods
    const versionMeta = document.createElement('meta')
    versionMeta.setAttribute('name', 'version')
    versionMeta.setAttribute('content', '3')
    rawHTMLHead.appendChild(versionMeta)

    const rawHTML = document.createElement('html')
    rawHTML.appendChild(rawHTMLHead)
    rawHTML.appendChild(rawHTMLBody)

    return {
      name: doc._id,
      content: rawHTML.outerHTML,
    }
  }

  return {
    serializeDocument,
    fieldFilter,
    languageObjectFieldFilter,
    serializeArray,
    serializeObject,
  }
}
