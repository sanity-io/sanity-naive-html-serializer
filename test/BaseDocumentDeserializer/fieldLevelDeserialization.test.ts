import {PortableTextBlock} from 'sanity'
import {fieldLevelArticle} from '../BaseDocumentSerializer/utils'
import {getDeserialized, toPlainText} from '../helpers'

test('Global test of working field-level functionality and snapshot match', () => {
  const deserialized = getDeserialized(fieldLevelArticle, 'field')
  expect(deserialized).toMatchSnapshot()
})

/*
 * Top-level plain text
 */

test('String and text types get deserialized correctly at top-level -- field level', () => {
  const deserialized = getDeserialized(fieldLevelArticle, 'field')
  expect(deserialized.title.en).toEqual(fieldLevelArticle.title.en)
  expect(deserialized.snippet.en).toEqual(fieldLevelArticle.snippet.en)
})

/*
 * Presence and accuracy of fields in "vanilla" deserialization -- objects
 */
test('Nested object contains accurate values -- field level', () => {
  const deserialized = getDeserialized(fieldLevelArticle, 'field')
  const origTitle = fieldLevelArticle.config.en.title
  const deserializedTitle = deserialized.config.en.title
  expect(origTitle).toEqual(deserializedTitle)

  const origBlockText = fieldLevelArticle.config.en.nestedArrayField
  const deserializedBlockText = deserialized.config.en.nestedArrayField

  const origKeys = origBlockText.map((block: PortableTextBlock) => block._key)
  const deserializedKeys = deserializedBlockText.map((block: PortableTextBlock) => block._key)

  expect(deserializedKeys.sort()).toEqual(origKeys.sort())
  expect(toPlainText(deserializedBlockText)).toEqual(toPlainText(origBlockText))
})

test('Nested object in an object contains accurate values -- field level', () => {
  const deserialized = getDeserialized(fieldLevelArticle, 'field')
  const origNestedObject = fieldLevelArticle.config.en.objectAsField
  const deserializedNestedObject = deserialized.config.en.objectAsField

  expect(origNestedObject.title).toEqual(deserializedNestedObject.title)

  const origKeys = origNestedObject.content.map((block: PortableTextBlock) => block._key)
  const deserializedKeys = deserializedNestedObject.content.map(
    (block: PortableTextBlock) => block._key
  )

  expect(origKeys.sort()).toEqual(deserializedKeys.sort())
  expect(toPlainText(deserializedNestedObject.content)).toEqual(
    toPlainText(origNestedObject.content)
  )
})

/*
 * Presence and accuracy of fields in vanilla deserialization -- arrays
 */

test('Array contains all serializable blocks with keys, in order -- field level', () => {
  const deserialized = getDeserialized(fieldLevelArticle, 'field')
  const origKeys = fieldLevelArticle.content.en.map((block: PortableTextBlock) => block._key)
  const deserializedKeys = deserialized.content.en.map((block: PortableTextBlock) => block._key)
  expect(deserializedKeys.sort()).toEqual(origKeys.sort())
})

test('Array contains top-level block text -- field level', () => {
  const deserialized = getDeserialized(fieldLevelArticle, 'field')
  expect(toPlainText(deserialized.content.en)).toEqual(toPlainText(fieldLevelArticle.content.en))
})

test('Object in array contains accurate values in nested object -- field level', () => {
  const deserialized = getDeserialized(fieldLevelArticle, 'field')
  const origTitle = fieldLevelArticle.content.en.find(
    (block: Record<string, any>) => block._type === 'objectField'
  ).objectAsField.title
  const deserializedTitle = deserialized.content.en.find(
    (block: Record<string, any>) => block._type === 'objectField'
  ).objectAsField.title
  expect(deserializedTitle).toEqual(origTitle)

  const origBlockText = toPlainText(
    fieldLevelArticle.content.en.find((block: Record<string, any>) => block._type === 'objectField')
      .objectAsField.content
  ).trim()
  const deserializedBlockText = toPlainText(
    fieldLevelArticle.content.en.find((block: Record<string, any>) => block._type === 'objectField')
      .objectAsField.content
  ).trim()
  expect(deserializedBlockText).toEqual(origBlockText)
})
