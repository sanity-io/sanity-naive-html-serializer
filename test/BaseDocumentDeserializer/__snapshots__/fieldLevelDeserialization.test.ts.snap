// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Global test of working field-level functionality and snapshot match 1`] = `
Object {
  "_id": "drafts.2947533e-1ea5-4116-955b-339608d3445d",
  "_rev": "l58oha-n06-1s4-f6i-74g6cuakl",
  "_type": "fieldLevelArticle",
  "config": Object {
    "en": Object {
      "_type": "objectField",
      "nestedArrayField": Array [
        Object {
          "_key": "f49b4d7e3e51",
          "_type": "block",
          "children": Array [
            Object {
              "_key": "randomKey-16",
              "_type": "span",
              "marks": Array [],
              "text": "This is block text 1 level deep",
            },
          ],
          "markDefs": Array [],
          "style": "normal",
        },
      ],
      "objectAsField": Object {
        "_type": "childObjectField",
        "content": Array [
          Object {
            "_key": "869168fe3a2a",
            "_type": "block",
            "children": Array [
              Object {
                "_key": "randomKey-14",
                "_type": "span",
                "marks": Array [],
                "text": "This is block text 2 levels deep",
              },
            ],
            "markDefs": Array [],
            "style": "normal",
          },
        ],
        "title": "This is one level deeper",
      },
      "title": "This is an object nested in a document",
    },
  },
  "content": Object {
    "en": Array [
      Object {
        "_key": "e2a39d768ff8",
        "_type": "block",
        "children": Array [
          Object {
            "_key": "randomKey-18",
            "_type": "span",
            "marks": Array [],
            "text": "This is block text at the top level.",
          },
        ],
        "markDefs": Array [],
        "style": "normal",
      },
      Object {
        "_key": "271b0c6ee984",
        "_type": "objectField",
        "objectAsField": Object {
          "_type": "childObjectField",
          "content": Array [
            Object {
              "_key": "4f24f6fbfae7",
              "_type": "block",
              "children": Array [
                Object {
                  "_key": "randomKey-24",
                  "_type": "span",
                  "marks": Array [],
                  "text": "This is block text in a nested object in an object in top-level block text.",
                },
              ],
              "markDefs": Array [],
              "style": "normal",
            },
          ],
          "title": "This is a nested object in an object in top-level block text.",
        },
        "title": "This is an object in top-level block text.",
      },
    ],
  },
  "snippet": Object {
    "en": "This is text in my text field",
  },
  "tags": Object {
    "en": Array [
      "tag 1",
      "tag 2",
      "tag 3",
    ],
  },
  "title": Object {
    "en": "My Field-Level Article",
  },
}
`;
